iroh-dns-server/src/dns.rs

```diff
@@ -0,0 +1,276 @@
+//! Implementation of a DNS name server for iroh node announces
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use bytes::Bytes;
+use hickory_server::{
+    authority::{Catalog, MessageResponse, ZoneType},
+    proto::{
+        self,
+        rr::{
+            rdata::{self},
+            RData, Record, RecordSet, RecordType, RrKey,
+        },
+        serialize::{binary::BinEncoder, txt::RDataParser},
+    },
+    resolver::Name,
+    server::{Request, RequestHandler, ResponseHandler, ResponseInfo},
+    store::in_memory::InMemoryAuthority,
+};
+
+use iroh_metrics::inc;
+use proto::{op::ResponseCode, rr::LowerName};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::BTreeMap,
+    io,
+    net::{Ipv4Addr, SocketAddrV4},
+    sync::Arc,
+    time::Duration,
+};
+use tokio::{
+    net::{TcpListener, UdpSocket},
+    sync::broadcast,
+};
+use tokio_util::sync::CancellationToken;
+use tracing::info;
+
+use self::node_authority::NodeAuthority;
+
+mod node_authority;
+use crate::config::Config;
+use crate::metrics::Metrics;
+use crate::store::SignedPacketStore;
+pub use node_authority::PacketSource;
+
+pub const DEFAULT_NS_TTL: u32 = 60 * 60 * 12; // 12h
+pub const DEFAULT_SOA_TTL: u32 = 60 * 60 * 24 * 14; // 14d
+pub const DEFAULT_A_TTL: u32 = 60 * 60; // 1h
+
+/// DNS server settings
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct DnsConfig {
+    /// The port to serve a local UDP DNS server at
+    pub port: u16,
+    /// SOA record data for any authoritative DNS records
+    pub default_soa: String,
+    /// Default time to live for returned DNS records (TXT & SOA)
+    pub default_ttl: u32,
+    /// Domain used for serving the `_iroh_node.<nodeid>.<origin>` DNS TXT entry
+    pub origin: String,
+    /// Domains where CNAME records will be set on `iroh_node.<nodeid>.origin`
+    pub additional_origins: Vec<String>,
+    pub ipv4_addr: Option<Ipv4Addr>,
+    pub ns_name: Option<String>,
+}
+
+pub async fn serve(
+    config: DnsConfig,
+    dns_server: DnsServer,
+    token: CancellationToken,
+) -> Result<()> {
+    const TCP_TIMEOUT: Duration = Duration::from_millis(1000);
+    let mut server = hickory_server::ServerFuture::new(dns_server);
+
+    let ip4_addr = config
+        .ipv4_addr
+        .unwrap_or_else(|| Ipv4Addr::new(127, 0, 0, 1));
+    let sock_addr = SocketAddrV4::new(ip4_addr, config.port);
+
+    server.register_socket(UdpSocket::bind(sock_addr).await?);
+    server.register_listener(
+        TcpListener::bind(sock_addr).await?,
+        TCP_TIMEOUT, // Duration::from_millis(settings.timeout_ms),
```

Comment:
what about this?

Reply:
I wasn't sure if we want the TCP timeout to be configurable, wdyt?

Reply:
seems okay for now to use a sensible default and if we need it make it configurable